server/src/main/java/org/opensearch/common/cache/stats/ImmutableCacheStatsHolder.java

```diff
@@ -23,15 +32,92 @@
+    private static final String SERIALIZATION_BEGIN_NODE = "_";
+    private static final String SERIALIZATION_DONE = "end";
+
+    ImmutableCacheStatsHolder(
+        CacheStatsHolder.Node originalStatsRoot,
+        List<String> levels,
+        List<String> originalDimensionNames,
+        String storeName
+    ) {
+        // Aggregate from the original CacheStatsHolder according to the levels passed in.
+        List<String> filteredLevels = filterLevels(levels, originalDimensionNames);
+        this.dimensionNames = filteredLevels; // The dimension names for this immutable snapshot should reflect the levels we aggregate in
+                                              // the snapshot
+        this.storeName = storeName;
+        this.statsRoot = aggregateByLevels(filteredLevels, originalStatsRoot, originalDimensionNames);
+    }
+
+    public ImmutableCacheStatsHolder(StreamInput in) throws IOException {
+        this.dimensionNames = List.of(in.readStringArray());
+        this.storeName = in.readString();
+        this.statsRoot = deserializeTree(in);
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeStringArray(dimensionNames.toArray(new String[0]));
+        out.writeString(storeName);
+        writeNode(statsRoot, out);
+        out.writeString(SERIALIZATION_DONE);
+    }
+
+    private void writeNode(Node node, StreamOutput out) throws IOException {
+        out.writeString(SERIALIZATION_BEGIN_NODE);
+        out.writeString(node.dimensionValue);
+        out.writeBoolean(node.children.isEmpty()); // Write whether this is a leaf node
+        node.stats.writeTo(out);
+
+        out.writeString(SERIALIZATION_CHILDREN_OPEN_BRACKET);
+        for (Map.Entry<String, Node> entry : node.children.entrySet()) {
+            out.writeString(entry.getKey());
+            writeNode(entry.getValue(), out);
+        }
+        out.writeString(SERIALIZATION_CHILDREN_CLOSE_BRACKET);
+    }
+
+    private Node deserializeTree(StreamInput in) throws IOException {
+        Stack<Node> stack = new Stack<>();
+        in.readString(); // Read and discard SERIALIZATION_BEGIN_NODE for the root node
+        Node statsRoot = readSingleNode(in);
+        Node current = statsRoot;
+        stack.push(statsRoot);
+        String nextSymbol = in.readString();
+        while (!nextSymbol.equals(SERIALIZATION_DONE)) {
+            switch (nextSymbol) {
+                case SERIALIZATION_CHILDREN_OPEN_BRACKET:
+                    stack.push(current);
+                    break;
+                case SERIALIZATION_CHILDREN_CLOSE_BRACKET:
+                    stack.pop();
+                    break;
+                case SERIALIZATION_BEGIN_NODE:
+                    current = readSingleNode(in);
+                    stack.peek().children.put(current.dimensionValue, current);
+            }
+            nextSymbol = in.readString();
+        }
+        return statsRoot;
```

Comment:
Thank you for this refactoring, the logic looks much intuitive and easier to follow now!